File: crates/parquet-lab/src/object_store.rs
//! A simulated object store: the subset of S3's semantics a Parquet reader depends on.
//!
//! Three operations, because a Parquet reader needs no more:
//!
//! - `HEAD` returns an object's size.
//! - `GET` with `Range: bytes=a-b` returns bytes `a` to `b` inclusive.
//! - `GET` with `Range: bytes=-n` returns the last `n` bytes, and the response says how large the
//!   whole object is. A reader can find a footer with no `HEAD` at all.
//!
//! [`MemoryStore`] holds the objects. [`TracingStore`] wraps any store, logs every request, and
//! charges each one a simulated cost from a [`NetworkModel`]. The trace the browser shows is
//! this log. Nothing is drawn that a request did not do.
//!
//! The costs are simulated on purpose. A real network would make every run of the book give
//! different numbers, and the book's rule is that every number it shows can be reproduced. The
//! model is deliberately simple: a fixed cost per request plus bytes over bandwidth. That
//! captures the property that shapes Parquet readers (a request is expensive, a byte is cheap)
//! and nothing else.

use std::collections::BTreeMap;
use std::fmt;

use crate::bytes::Span;

/// Which bytes a `GET` asks for.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum GetRange {
    /// `bytes=start-(end-1)`: a span of absolute offsets.
    Bounded(Span),
    /// `bytes=-n`: the last `n` bytes, whatever the object's size.
    Suffix(u64),
}

impl GetRange {
    /// The value of the HTTP `Range` header for this request.
    pub fn header(&self) -> String {
        match self {
            GetRange::Bounded(span) => span.http_range(),
            GetRange::Suffix(n) => format!("bytes=-{n}"),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum StoreError {
    NotFound {
        key: String,
    },
    /// HTTP 416: the range starts at or after the end of the object.
    RangeNotSatisfiable {
        key: String,
        range: String,
        size: u64,
    },
}

impl fmt::Display for StoreError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            StoreError::NotFound { key } => write!(f, "404: no object named {key}"),
            StoreError::RangeNotSatisfiable { key, range, size } => {
                write!(f, "416: {range} is outside {key}, which is {size} bytes")
            }
        }
    }
}

impl std::error::Error for StoreError {}

/// The body of a successful `GET`.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct GetResult {
    pub bytes: Vec<u8>,
    /// Which bytes of the object these are: the response's `Content-Range`.
    pub span: Span,
    /// The object's total size, which `Content-Range` also carries.
    pub object_size: u64,
}

/// What a Parquet reader needs from storage.
///
/// `why` is not part of any storage protocol. It is instrumentation: the reader says what it
/// wants the bytes for, the trace records it, and the browser can then answer "why did the
/// reader request these bytes?" from the reader's own words rather than from a caption.
pub trait ObjectStore {
    fn head(&mut self, key: &str, why: &str) -> Result<u64, StoreError>;
    fn get(&mut self, key: &str, range: GetRange, why: &str) -> Result<GetResult, StoreError>;
}

/// Objects held in memory: the book's fixtures, embedded in the page or read from disk.
#[derive(Clone, Debug, Default)]
pub struct MemoryStore {
    objects: BTreeMap<String, Vec<u8>>,
}

impl MemoryStore {
    pub fn new() -> MemoryStore {
        MemoryStore::default()
    }

    pub fn put(&mut self, key: &str, bytes: Vec<u8>) {
        self.objects.insert(key.to_string(), bytes);
    }

    fn object(&self, key: &str) -> Result<&Vec<u8>, StoreError> {
        self.objects.get(key).ok_or_else(|| StoreError::NotFound {
            key: key.to_string(),
        })
    }
}

impl ObjectStore for MemoryStore {
    fn head(&mut self, key: &str, _why: &str) -> Result<u64, StoreError> {
        Ok(self.object(key)?.len() as u64)
    }

    /// Serve a range the way S3 does: a range that runs past the end is cut short rather than
    /// refused, a suffix longer than the object returns the whole object, and only a range that
    /// starts at or beyond the end is an error.
    fn get(&mut self, key: &str, range: GetRange, _why: &str) -> Result<GetResult, StoreError> {
        let data = self.object(key)?;
        let size = data.len() as u64;
        let span = match range {
            GetRange::Bounded(s) => {
                if s.start >= size || s.is_empty() {
                    return Err(StoreError::RangeNotSatisfiable {
                        key: key.to_string(),
                        range: range.header(),
                        size,
                    });
                }
                Span::new(s.start, s.end.min(size))
            }
            GetRange::Suffix(n) => Span::new(size.saturating_sub(n), size),
        };
        Ok(GetResult {
            bytes: data[span.start as usize..span.end as usize].to_vec(),
            span,
            object_size: size,
        })
    }
}

/// What a request costs: a fixed latency, and the time to move its bytes.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct NetworkModel {
    /// Time from sending a request to the first byte of the response, in microseconds.
    pub latency_us: u64,
    /// Bytes per second once the response is flowing. Zero means unlimited.
    pub bandwidth_bytes_per_sec: u64,
}

impl NetworkModel {
    /// Microseconds to complete one request that returns `bytes` bytes.
    pub fn cost_us(&self, bytes: u64) -> u64 {
        let transfer = match self.bandwidth_bytes_per_sec {
            0 => 0,
            bw => (bytes * 1_000_000).div_ceil(bw),
        };
        self.latency_us + transfer
    }
}

impl Default for NetworkModel {
    /// A request that takes a while to start and then moves bytes quickly: the shape of object
    /// storage over a network. The two numbers are a scenario, not a measurement; the browser
    /// lets the reader change both.
    fn default() -> Self {
        NetworkModel {
            latency_us: 20_000,
            bandwidth_bytes_per_sec: 100_000_000,
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Method {
    Head,
    Get,
}

impl Method {
    pub fn name(&self) -> &'static str {
        match self {
            Method::Head => "HEAD",
            Method::Get => "GET",
        }
    }
}

/// One entry in the trace: a request, what came back, and what it cost.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Request {
    pub seq: usize,
    pub method: Method,
    pub key: String,
    /// The `Range` header, if the request had one.
    pub range: Option<String>,
    pub why: String,
    /// 200 for a `HEAD`, 206 for a satisfied range, or the error status.
    pub status: u16,
    /// The bytes that came back, as offsets into the object.
    pub returned: Option<Span>,
    pub bytes_returned: u64,
    pub start_us: u64,
    pub end_us: u64,
}

/// Wraps a store, and records and prices every request that passes through it.
///
/// Requests are issued one after another, so each starts when the previous one ended. A
/// reader that issues requests concurrently is a later chapter's subject (ch10), and when it
/// arrives it will change this clock, not the requests.
#[derive(Debug)]
pub struct TracingStore<S> {
    inner: S,
    pub model: NetworkModel,
    pub requests: Vec<Request>,
    clock_us: u64,
}

impl<S: ObjectStore> TracingStore<S> {
    pub fn new(inner: S, model: NetworkModel) -> TracingStore<S> {
        TracingStore {
            inner,
            model,
            requests: Vec::new(),
            clock_us: 0,
        }
    }

    /// Total simulated time for every request so far.
    pub fn elapsed_us(&self) -> u64 {
        self.clock_us
    }

    pub fn bytes_returned(&self) -> u64 {
        self.requests.iter().map(|r| r.bytes_returned).sum()
    }

    fn record(
        &mut self,
        method: Method,
        key: &str,
        range: Option<String>,
        why: &str,
        status: u16,
        returned: Option<Span>,
    ) {
        let bytes = returned.map(|s| s.len()).unwrap_or(0);
        let start_us = self.clock_us;
        self.clock_us += self.model.cost_us(bytes);
        self.requests.push(Request {
            seq: self.requests.len() + 1,
            method,
            key: key.to_string(),
            range,
            why: why.to_string(),
            status,
            returned,
            bytes_returned: bytes,
            start_us,
            end_us: self.clock_us,
        });
    }
}

fn status_of(e: &StoreError) -> u16 {
    match e {
        StoreError::NotFound { .. } => 404,
        StoreError::RangeNotSatisfiable { .. } => 416,
    }
}

impl<S: ObjectStore> ObjectStore for TracingStore<S> {
    fn head(&mut self, key: &str, why: &str) -> Result<u64, StoreError> {
        let result = self.inner.head(key, why);
        let status = result.as_ref().map(|_| 200).unwrap_or_else(status_of);
        self.record(Method::Head, key, None, why, status, None);
        result
    }

    fn get(&mut self, key: &str, range: GetRange, why: &str) -> Result<GetResult, StoreError> {
        let result = self.inner.get(key, range, why);
        let (status, returned) = match &result {
            Ok(r) => (206, Some(r.span)),
            Err(e) => (status_of(e), None),
        };
        self.record(
            Method::Get,
            key,
            Some(range.header()),
            why,
            status,
            returned,
        );
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn store() -> MemoryStore {
        let mut s = MemoryStore::new();
        s.put("data.parquet", (0u8..100).collect());
        s
    }

    #[test]
    fn a_bounded_range_returns_exactly_those_bytes() {
        let r = store()
            .get("data.parquet", GetRange::Bounded(Span::new(10, 14)), "")
            .unwrap();
        assert_eq!(r.bytes, [10, 11, 12, 13]);
        assert_eq!(r.span, Span::new(10, 14));
        assert_eq!(r.object_size, 100);
    }

    #[test]
    fn a_range_past_the_end_is_cut_short_not_refused() {
        let r = store()
            .get("data.parquet", GetRange::Bounded(Span::new(95, 200)), "")
            .unwrap();
        assert_eq!(r.span, Span::new(95, 100));
    }

    #[test]
    fn a_range_starting_past_the_end_is_416() {
        let e = store()
            .get("data.parquet", GetRange::Bounded(Span::new(100, 101)), "")
            .unwrap_err();
        assert!(matches!(
            e,
            StoreError::RangeNotSatisfiable { size: 100, .. }
        ));
    }

    #[test]
    fn a_suffix_range_needs_no_size_and_reports_it() {
        let r = store()
            .get("data.parquet", GetRange::Suffix(8), "")
            .unwrap();
        assert_eq!(r.span, Span::new(92, 100));
        assert_eq!(r.object_size, 100);
        let whole = store()
            .get("data.parquet", GetRange::Suffix(1000), "")
            .unwrap();
        assert_eq!(whole.span, Span::new(0, 100));
    }

    #[test]
    fn the_trace_records_every_request_and_its_cost() {
        let model = NetworkModel {
            latency_us: 1000,
            bandwidth_bytes_per_sec: 1_000_000,
        };
        let mut t = TracingStore::new(store(), model);
        t.head("data.parquet", "size").unwrap();
        t.get("data.parquet", GetRange::Suffix(8), "tail").unwrap();
        let _ = t.get("missing", GetRange::Suffix(8), "oops");
        assert_eq!(t.requests.len(), 3);
        assert_eq!(t.requests[0].method, Method::Head);
        assert_eq!(t.requests[1].range.as_deref(), Some("bytes=-8"));
        assert_eq!(t.requests[1].returned, Some(Span::new(92, 100)));
        assert_eq!(t.requests[2].status, 404);
        // 1000us per request, plus 8 bytes at a byte per microsecond for the GET.
        assert_eq!(t.requests[1].start_us, 1000);
        assert_eq!(t.requests[1].end_us, 2008);
        assert_eq!(t.elapsed_us(), 3008);
        assert_eq!(t.bytes_returned(), 8);
    }

    #[test]
    fn unlimited_bandwidth_costs_only_latency() {
        let m = NetworkModel {
            latency_us: 5,
            bandwidth_bytes_per_sec: 0,
        };
        assert_eq!(m.cost_us(1 << 30), 5);
    }
}

File: python/parquet_lab/object_store.py
"""A simulated object store: the subset of S3's semantics a Parquet reader depends on.

Three operations, because a Parquet reader needs no more:

- ``HEAD`` returns an object's size.
- ``GET`` with ``Range: bytes=a-b`` returns bytes ``a`` to ``b`` inclusive.
- ``GET`` with ``Range: bytes=-n`` returns the last ``n`` bytes, and the response says how large
  the whole object is. A reader can find a footer with no ``HEAD`` at all.

:class:`MemoryStore` holds the objects. :class:`TracingStore` wraps any store, logs every
request, and charges each one a simulated cost from a :class:`NetworkModel`. The trace the
browser shows is this log. Nothing is drawn that a request did not do.

The costs are simulated on purpose. A real network would make every run of the book give
different numbers, and the book's rule is that every number it shows can be reproduced. The
model is deliberately simple: a fixed cost per request plus bytes over bandwidth. That captures
the property that shapes Parquet readers (a request is expensive, a byte is cheap) and nothing
else.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from .bytes import Span


@dataclass(frozen=True)
class Bounded:
    """``bytes=start-(end-1)``: a span of absolute offsets."""

    span: Span

    def header(self) -> str:
        return self.span.http_range()


@dataclass(frozen=True)
class Suffix:
    """``bytes=-n``: the last ``n`` bytes, whatever the object's size."""

    n: int

    def header(self) -> str:
        return f"bytes=-{self.n}"


@dataclass(frozen=True)
class All:
    """No ``Range`` header: the whole object (ch14)."""

    def header(self) -> str:
        return "the whole object"


GetRange = Bounded | Suffix | All
"""Which bytes a ``GET`` asks for."""


class StoreError(Exception):
    status = 0


class NotFound(StoreError):
    status = 404

    def __init__(self, key: str) -> None:
        super().__init__(f"404: no object named {key}")


class RangeNotSatisfiable(StoreError):
    """HTTP 416: the range starts at or after the end of the object."""

    status = 416

    def __init__(self, key: str, range: str, size: int) -> None:
        super().__init__(f"416: {range} is outside {key}, which is {size} bytes")


@dataclass
class GetResult:
    """The body of a successful ``GET``."""

    data: bytes
    span: Span
    """Which bytes of the object these are: the response's ``Content-Range``."""
    object_size: int
    """The object's total size, which ``Content-Range`` also carries."""


class ObjectStore(Protocol):
    """What a Parquet reader needs from storage.

    ``why`` is not part of any storage protocol. It is instrumentation: the reader says what it
    wants the bytes for, the trace records it, and the browser can then answer "why did the
    reader request these bytes?" from the reader's own words rather than from a caption.
    """

    def head(self, key: str, why: str) -> int: ...

    def get(self, key: str, range: GetRange, why: str) -> GetResult: ...

    def list(self, prefix: str, why: str) -> list[tuple[str, int]]:
        """The keys that start with ``prefix``, in order, with each object's size: S3's
        ``LIST`` (ch14)."""
        ...

    def next_phase(self) -> None:
        """Say that the next request depends on the answers to earlier ones, so it cannot start
        until they have finished. A store that does not model time ignores it."""


class MemoryStore:
    """Objects held in memory: the book's fixtures, embedded in the page or read from disk."""

    def __init__(self) -> None:
        self.objects: dict[str, bytes] = {}

    def put(self, key: str, data: bytes) -> None:
        self.objects[key] = bytes(data)

    def _object(self, key: str) -> bytes:
        if key not in self.objects:
            raise NotFound(key)
        return self.objects[key]

    def list(self, prefix: str, why: str) -> list[tuple[str, int]]:
        return [(k, len(v)) for k, v in sorted(self.objects.items()) if k.startswith(prefix)]

    def head(self, key: str, why: str) -> int:
        return len(self._object(key))

    def get(self, key: str, range: GetRange, why: str) -> GetResult:
        """Serve a range the way S3 does: a range that runs past the end is cut short rather than
        refused, a suffix longer than the object returns the whole object, and only a range that
        starts at or beyond the end is an error."""
        data = self._object(key)
        size = len(data)
        if isinstance(range, Bounded):
            s = range.span
            if s.start >= size or s.length == 0:
                raise RangeNotSatisfiable(key, range.header(), size)
            span = Span(s.start, min(s.end, size))
        elif isinstance(range, Suffix):
            span = Span(max(size - range.n, 0), size)
        else:
            span = Span(0, size)
        return GetResult(data[span.start : span.end], span, size)

    def next_phase(self) -> None:
        pass


@dataclass(frozen=True)
class NetworkModel:
    """What a request costs: a fixed latency, and the time to move its bytes.

    The defaults are a request that takes a while to start and then moves bytes quickly: the
    shape of object storage over a network. The two numbers are a scenario, not a measurement;
    the browser lets the reader change both.
    """

    latency_us: int = 20_000
    """Time from sending a request to the first byte of the response, in microseconds."""
    bandwidth_bytes_per_sec: int = 100_000_000
    """Bytes per second once the response is flowing. Zero means unlimited."""

    def cost_us(self, nbytes: int) -> int:
        """Microseconds to complete one request that returns ``nbytes`` bytes."""
        bw = self.bandwidth_bytes_per_sec
        transfer = 0 if bw == 0 else -(-nbytes * 1_000_000 // bw)
        return self.latency_us + transfer


@dataclass
class Request:
    """One entry in the trace: a request, what came back, and what it cost."""

    seq: int
    method: str
    key: str
    range: str | None
    """The ``Range`` header, if the request had one."""
    why: str
    status: int
    """200 for a ``HEAD``, 206 for a satisfied range, or the error status."""
    returned: Span | None
    """The bytes that came back, as offsets into the object."""
    bytes_returned: int
    start_us: int
    end_us: int
    connection: int
    """Which connection carried it, from 0."""
    phase: int
    """Which phase it belonged to, from 0."""


class TracingStore:
    """Wraps a store, and records and prices every request that passes through it.

    The store has ``connections`` connections, one by default. A request goes on whichever
    connection is free first, and starts when that connection is free and the current phase has
    begun. With one connection every request starts when the previous one ended. With several,
    requests in the same phase overlap in time (ch10).

    A **phase** is a set of requests that do not depend on each other. A request that needs the
    answer to an earlier one, such as a data read that needs the footer, belongs to a later
    phase: :meth:`next_phase` makes every later request wait for all earlier ones.
    """

    def __init__(self, inner: ObjectStore, model: NetworkModel, connections: int = 1) -> None:
        self.inner, self.model = inner, model
        self.requests: list[Request] = []
        self.free_at = [0] * max(connections, 1)
        """When each connection is next free."""
        self.phase_start = 0
        """No request starts before this: the end of the previous phase."""
        self.phase = 0

    def elapsed_us(self) -> int:
        """Total simulated time: when the last request finished."""
        return max((r.end_us for r in self.requests), default=0)

    def bytes_returned(self) -> int:
        return sum(r.bytes_returned for r in self.requests)

    def _record(self, method, key, range, why, status, returned) -> None:
        nbytes = returned.length if returned else 0
        # The connection that is free first; ties go to the lowest number.
        free, connection = min((t, i) for i, t in enumerate(self.free_at))
        start_us = max(free, self.phase_start)
        end_us = start_us + self.model.cost_us(nbytes)
        self.free_at[connection] = end_us
        self.requests.append(
            Request(
                seq=len(self.requests) + 1,
                method=method,
                key=key,
                range=range,
                why=why,
                status=status,
                returned=returned,
                bytes_returned=nbytes,
                start_us=start_us,
                end_us=end_us,
                connection=connection,
                phase=self.phase,
            )
        )

    def next_phase(self) -> None:
        """Start a new phase: later requests wait until every request so far has finished."""
        self.phase_start = self.elapsed_us()
        self.phase += 1

    def list(self, prefix: str, why: str) -> list[tuple[str, int]]:
        """A ``LIST`` is one request. Its response, a page of keys, is not counted in bytes: the
        model prices the request, not the listing's size."""
        out = self.inner.list(prefix, why)
        self._record("LIST", prefix, None, why, 200, None)
        return out

    def head(self, key: str, why: str) -> int:
        try:
            size = self.inner.head(key, why)
        except StoreError as e:
            self._record("HEAD", key, None, why, e.status, None)
            raise
        self._record("HEAD", key, None, why, 200, None)
        return size

    def get(self, key: str, range: GetRange, why: str) -> GetResult:
        header = None if isinstance(range, All) else range.header()
        try:
            result = self.inner.get(key, range, why)
        except StoreError as e:
            self._record("GET", key, header, why, e.status, None)
            raise
        self._record("GET", key, header, why, 206, result.span)
        return result
